### PR TITLE
Fix W.splitFileName "/\\?/a:"

### LINKED ISSUE #219: possible reversion of W.splitFileName "/\\?/a:"
Before 1.4.100.2, `System.FilePath.Windows.splitFileName "/\\?/a:" == ("/\\?/a:","")` and in that version and since it changed to `("/\\?/","a:")`

I don't fully understand Windows path semantics, and this path was generated by QuickCheck, so I don't know which is right, but 1.4.100.2 did not document any change besides optimizations. Which makes me suspect this is an unintentional change.

Also, https://github.com/haskell/filepath/pull/189/commits/6a1c98dc5eb8021107d59903bbbb60137a6f9dbd fixed a reversion in `W.splitFileName "\\\\?\\A:\\fred"` which seems very similar to this.

```
ghci> System.FilePath.Windows.splitFileName "\\\\?\\A:\\fred"
("\\\\?\\A:\\","fred")
ghci> System.FilePath.Windows.splitFileName "\\\\?\\A:\\"
("\\\\?\\A:\\","")
ghci> System.FilePath.Windows.splitFileName "\\\\?\\A:"
("\\\\?\\","A:")
```

### PR DESCRIPTION
**Pull request by Bodigrim**
Fixes #219.

### CODE REVIEW

File: System/FilePath/Internal.hs [hunk 1]
```diff
@@ -644,15 +645,26 @@ splitFileName_ fp
...
+      -- This handles inputs like "//?/A:" and "//?/A:foo"
+      | isPathSeparator s1
+      , isPathSeparator s2
+      , Just (s3, s4, bs'') <- uncons2 bs'
+      , s3 == _question
+      , isPathSeparator s4
+      , null bs''
+      , Just (drive, rest) <- readDriveLetter file
+      -> (dirSlash <> drive, rest)
+    _ -> (dirSlash, file)
```
Comment (hasufell) [inline System/FilePath/Internal.hs:665]: I think this line is redundant? I'tll fall through to `| otherwise`, no?

**Comment by hasufell** [comment]
We have equivalence tests that compare this function with the old implementation. It's depressing that it wasn't caused. The quickcheck generators are not good enough and don't produce enough "odd data".

I think one way forward is to implement the ABNF that I have here: https://github.com/hasufell/filepath/blob/16e5374620189b27eca1eed09642ec02b2222fc8/System/FilePath/Internal/Parser.hs#L157-L165

...at least for the test suite and use that to generate windows filepaths.

File: System/FilePath/Internal.hs [hunk 2]
```diff
@@ -644,15 +645,26 @@ splitFileName_ fp
...
-  = (fp, mempty)
+  = case uncons2 dirSlash of
+    Just (s1, s2, bs')
+      | isPathSeparator s1
+      -- If bs' is empty, then s2 as the last character of dirSlash must be a path separator,
+      -- so we are in the middle of shared drive.
+      -- Otherwise, since s1 is a path separator, we might be in the middle of UNC path.
+      , null bs' || maybe False isIncompleteUNC (readDriveUNC dirSlash)
+      -> (fp, mempty)
+      -- This handles inputs like "//?/A:" and "//?/A:foo"
```
Comment (hasufell) [inline System/FilePath/Internal.hs:656]: Trivia: `//?/A:foo` is actually invalid in some sense... it's hard to interpret it. UNC paths are absolute, but `A:foo` on its own is a relative directory (relative to current working dir on drive `A`). The windows API probably won't care, but no such file can ever exist, I think.

File: System/FilePath/Internal.hs (hunk 1 above)
Comment (Bodigrim) [reply System/FilePath/Internal.hs:665]: Unfortunately it does not fall through, the line is required.

File: System/FilePath/Internal.hs (hunk 2 above)
Comment (Bodigrim) [reply System/FilePath/Internal.hs:656]: True, but `filepath-1.4.2.2` does not mind to process `//?/A:foo` even if it's complete nonsense.

**Comment by Bodigrim** [comment]
I started this PR with thinking about properties of `splitFileName`. One of them is that `takeDrive x `isPrefixOf` dropFileName x`. Testing such property indeed reveals the originally reported issue with `\\?\a:` and also `\\?\a:foo`.

File: System/FilePath/Internal.hs (hunk 1 above)
Comment (hasufell) [reply System/FilePath/Internal.hs:665]: Oops, I misread the code.

**Comment by hasufell** [comment]
I pushed updated equivalence tests that now make use of the ABNF to generate more odd cases. Lots of new failures:

```
$ cabal run --enable-tests filepath-equivalent-tests -- --quickcheck-tests 50_000 -p 'windows'
equivalence
  pathSeparator (windows):             OK
    +++ OK, passed 1 test.
  pathSeparators (windows):            OK
    +++ OK, passed 1 test.
  isPathSeparator (windows):           OK (0.14s)
    +++ OK, passed 50000 tests.
  searchPathSeparator (windows):       OK
    +++ OK, passed 1 test.
  isSearchPathSeparator (windows):     OK (0.14s)
    +++ OK, passed 50000 tests.
  extSeparator (windows):              OK
    +++ OK, passed 1 test.
  isExtSeparator (windows):            OK (0.12s)
    +++ OK, passed 50000 tests.
  splitSearchPath (windows):           OK (3.64s)
    +++ OK, passed 50000 tests.
  splitExtension (windows):            OK (0.68s)
    +++ OK, passed 50000 tests.
  takeExtension (windows):             OK (0.62s)
    +++ OK, passed 50000 tests.
  replaceExtension (windows):          OK (1.08s)
    +++ OK, passed 50000 tests.
  dropExtension (windows):             OK (0.67s)
    +++ OK, passed 50000 tests.
  addExtension (windows):              OK (1.00s)
    +++ OK, passed 50000 tests.
  hasExtension (windows):              FAIL (0.16s)
    *** Failed! Falsified (after 15339 tests and 1 shrink):
    NS FileNameSpace [] (NST1 'a' UnixSep":|"[WindowsSep,WindowsSep,UnixSep,WindowsSep,WindowsSep,UnixSep,UnixSep,UnixSep,UnixSep,UnixSep,UnixSep,WindowsSep] (Rel2 FileName (NonEmptyString 'f'":|""") (Just DS2 NonEmptyString '.'":|"""))) (\\?\a:/\\/\\//////\fJust (DS2 (NonEmptyString ('.' :| ""))))
    Use --quickcheck-replay=274593 to reproduce.
    Use -p '/windows/&&/hasExtension (windows)/' to rerun this test only.
  splitExtensions (windows):           OK (0.82s)
    +++ OK, passed 50000 tests.
  dropExtensions (windows):            FAIL (0.41s)
    *** Failed! Falsified (after 25799 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' UnixSep":|"[UnixSep,WindowsSep,UnixSep,UnixSep,UnixSep] (Rel2 FileName (NonEmptyString 'G'":|""U\16436\1071014W\STX^.\DC3\139778e `1\EM\FS`") Nothing)) (\\?\a://\///GU䀴􅞦W^.𢈂e `1`Nothing)
    Use --quickcheck-replay=374287 to reproduce.
    Use -p '/windows/&&/dropExtensions (windows)/' to rerun this test only.
  takeExtensions (windows):            OK (0.40s)
    +++ OK, passed 50000 tests.
  replaceExtensions (windows):         OK (1.23s)
    +++ OK, passed 50000 tests.
  isExtensionOf (windows):             OK (0.48s)
    +++ OK, passed 50000 tests.
  stripExtension (windows):            OK (0.50s)
    +++ OK, passed 50000 tests.
  splitFileName (windows):             FAIL
    *** Failed! Falsified (after 1251 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' UnixSep":|"[WindowsSep,UnixSep,WindowsSep,WindowsSep,WindowsSep,UnixSep,WindowsSep] (Rel2 FileName (NonEmptyString '\169168'":|""\US") Nothing)) (\\?\a:/\/\\\/\𩓐Nothing)
    Use --quickcheck-replay=139158 to reproduce.
    Use -p '/windows/&&/splitFileName (windows)/' to rerun this test only.
  takeFileName (windows):              FAIL (0.12s)
    *** Failed! Falsified (after 11856 tests and 1 shrink):
    NS FileNameSpace [] (NST1 'a' WindowsSep":|"[UnixSep,UnixSep,WindowsSep,UnixSep,WindowsSep,UnixSep] (Rel2 FileName (NonEmptyString 'w'":|""") Nothing)) (\\?\a:\//\/\/wNothing)
    Use --quickcheck-replay=287678 to reproduce.
    Use -p '/windows/&&/takeFileName (windows)/' to rerun this test only.
  replaceFileName (windows):           FAIL (0.11s)
    *** Failed! Falsified (after 5399 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' WindowsSep":|"[UnixSep,WindowsSep,WindowsSep,UnixSep,UnixSep,WindowsSep,WindowsSep,WindowsSep] (Rel2 FileName (NonEmptyString 'Z'":|""g!k\186849]m0\1113081[)") Nothing)) (\\?\a:\/\\//\\\Zg!k𭧡]m0􏯹[)Nothing)
    ""
    Use --quickcheck-replay=761264 to reproduce.
    Use -p '/windows/&&/replaceFileName (windows)/' to rerun this test only.
  dropFileName (windows):              FAIL
    *** Failed! Falsified (after 331 tests and 1 shrink):
    NS FileNameSpace [] (NST1 'a' WindowsSep":|"[UnixSep,UnixSep,WindowsSep] (Rel2 FileName (NonEmptyString '\CAN'":|"")?\1053985\1099684\&0E") Nothing)) (\\?\a:\//\)?􁔡􌞤0ENothing)
    Use --quickcheck-replay=759365 to reproduce.
    Use -p '/windows/&&/dropFileName (windows)/' to rerun this test only.
  takeBaseName (windows):              FAIL
    *** Failed! Falsified (after 731 tests and 1 shrink):
    NS FileNameSpace [] (NST1 'a' WindowsSep":|"[UnixSep,WindowsSep,WindowsSep,UnixSep,WindowsSep,WindowsSep,UnixSep,WindowsSep] (Rel2 FileName (NonEmptyString '@'":|""@\1089231\999202") Nothing)) (\\?\a:\/\\/\\/\@@􉻏󳼢Nothing)
    Use --quickcheck-replay=191303 to reproduce.
    Use -p '/windows/&&/takeBaseName (windows)/' to rerun this test only.
  replaceBaseName (windows):           FAIL (0.02s)
    *** Failed! Falsified (after 668 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' UnixSep":|"[] (Rel2 FileName (NonEmptyString '/'":|""") Nothing)) (\\?\a://Nothing)
    ""
    Use --quickcheck-replay=604486 to reproduce.
    Use -p '/windows/&&/replaceBaseName (windows)/' to rerun this test only.
  takeDirectory (windows):             FAIL (0.07s)
    *** Failed! Falsified (after 4487 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' WindowsSep":|"[] (Rel2 FileName (NonEmptyString '/'":|""\1074409\1004589{\DC2b") Nothing)) (\\?\a:\/􆓩󵐭{bNothing)
    Use --quickcheck-replay=128258 to reproduce.
    Use -p '/windows/&&/takeDirectory (windows)/' to rerun this test only.
  replaceDirectory (windows):          FAIL
    *** Failed! Falsified (after 419 tests and 2 shrinks):
    NS FileNameSpace [] (NST1 'a' UnixSep":|"[WindowsSep,WindowsSep,WindowsSep,UnixSep,UnixSep,UnixSep,UnixSep,UnixSep,WindowsSep,UnixSep,WindowsSep,WindowsSep,WindowsSep] (Rel2 FileName (NonEmptyString '\1112743'":|""6\NUL") Nothing)) (\\?\a:/\\\/////\/\\\􏪧6Nothing)
    ""
    Use --quickcheck-replay=886203 to reproduce.
    Use -p '/windows/&&/replaceDirectory (windows)/' to rerun this test only.
  combine (windows):                   OK (0.91s)
    +++ OK, passed 50000 tests.
  splitPath (windows):                 OK (0.96s)
    +++ OK, passed 50000 tests.
  joinPath (windows):                  OK (0.77s)
    +++ OK, passed 50000 tests.
  splitDirectories (windows):          OK (1.11s)
    +++ OK, passed 50000 tests.
  splitDrive (windows):                OK (0.61s)
    +++ OK, passed 50000 tests.
  joinDrive (windows):                 OK (0.90s)
    +++ OK, passed 50000 tests.
  takeDrive (windows):                 OK (0.23s)
    +++ OK, passed 50000 tests.
  hasDrive (windows):                  OK (0.18s)
    +++ OK, passed 50000 tests.
  dropDrive (windows):                 OK (0.58s)
    +++ OK, passed 50000 tests.
  isDrive (windows):                   OK (0.23s)
    +++ OK, passed 50000 tests.
  hasTrailingPathSeparator (windows):  OK (0.37s)
    +++ OK, passed 50000 tests.
  addTrailingPathSeparator (windows):  OK (0.62s)
    +++ OK, passed 50000 tests.
  dropTrailingPathSeparator (windows): OK (0.60s)
    +++ OK, passed 50000 tests.
  normalise (windows):                 OK (1.37s)
    +++ OK, passed 50000 tests.
  equalFilePath (windows):             OK (1.29s)
    +++ OK, passed 50000 tests.
  makeRelative (windows):              OK (1.90s)
    +++ OK, passed 50000 tests.
  isRelative (windows):                OK (0.16s)
    +++ OK, passed 50000 tests.
  isAbsolute (windows):                OK (0.14s)
    +++ OK, passed 50000 tests.
  isValid (windows):                   OK (0.38s)
    +++ OK, passed 50000 tests.
  makeValid (windows):                 OK (1.01s)
    +++ OK, passed 50000 tests.

10 out of 48 tests failed (26.88s)
```

**Comment by hasufell** [comment]
More specifically for `splitFileName`:

This PR:

```
ghci> System.FilePath.Windows.splitFileName "\\\\?\\a:///\\\\\\\\k1"
("\\\\?\\a:///\\\\\\\\k1","")
```

filepath-1.4.2.2:

```
ghci> System.FilePath.Windows.splitFileName "\\\\?\\a:///\\\\\\\\k1"
("\\\\?\\a:///\\\\\\\\","k1")
```

**Comment by hasufell** [comment]
The bug was in `hasPenultimateColon`

File: System/FilePath/Internal.hs [hunk 3]
```diff
@@ -644,20 +646,46 @@ splitFileName_ fp
...
+      | isPathSeparator s1
+      , isPathSeparator s2
+      , Just (s3, s4, bs'') <- uncons2 bs'
+      , s3 == _question
```
Comment (hasufell) [inline System/FilePath/Internal.hs:661]: Note that this is in fact a bug (although the old filepath behaves the same way):

```
ghci> System.FilePath.Windows.splitFileName "\\\\.\\a:foo"
("\\\\.\\","a:foo")
```

There are three UNC namespaces:

* file namespace: `//?/`
* device namespace: `//./`
* NT namespace `/??/`

I'm still unsure whether fixing those bugs will cause more good than harm, but those are bugs.

**Pull request merged**